### DISCUSSION TITLE
stream a file

### DISCUSSION THREAD

**Discussion by duckcantcode** [dokan-dev/dokany#1181] [Q&A] [answered]
Hi! I'm making a project in Rust with Dokan that I want to show the user placeholder audio files, and when you open one, the actual music/data is streamed from a remote server. Is it possible to do this?

**Comment by Liryna** [comment]
Hi @duckfromdiscord ,

Yes you can! The filesystem content and what is read is up to the implementation.

**Reply by duckcantcode** [reply]
Awesome. Is there some way to avoid extremely long loading times? Right now I have all file reads directed to a HTTP request reading at the offset with the length of the buffer, and for some reason, perhaps windows trying to load metadata, it can take multiple minutes for a single folder to load.

Also:
At the moment, for file create calls, I am asking the server for a link to a static audio file, and that is what allows me to do this, since I can query the file itself and give Windows that file size. But I would like to implement an option to tell the server to use downscaled audio, which means it's being streamed and there is no definitive file size since we're basically getting the raw output of an `ffmpeg` command being run serverside. I can figure out how to get the audio stream in byte form over HTTP, I am already using a crate for this, but I am not sure what size I should give Windows so that it will read the file in its entirety.

Basically, I don't know how to deal with providing a stream of bytes as an actual file to Dokan.

**Reply by Liryna** [reply]
> Is there some way to avoid extremely long loading times?

This would require a local caching adapted to your server API. Windows does not need the whole file to be downloaded to play it.

> there is no definitive file size since we're basically getting the raw output of an ffmpeg command being run serverside

That could indeed be challenging as you need to provide the size beforehand. I don't have an answer 😢 sorry.

**Reply by duckcantcode** [reply]
> This would require a local caching adapted to your server API. Windows does not need the whole file to be downloaded to play it.

I was thinking of perhaps caching the first 65535 or 65535*2 bytes of a file locally or in memory, since through my debugging I have found that Windows seems to make one or two requests into a buffer of length 65535 for each file when a folder is opened.

> That could indeed be challenging as you need to provide the size beforehand. I don't have an answer 😢 sorry.

That's fine, it is equally important to know that there is not an obvious answer I have missed. Thank you very much for your help. Is it possible to simply change the file size provided as more data is streamed or will this cause Windows to abruptly stop at the first size given? Or, conversely, is it possible to provide a very large file size and stop sending data in the middle of it when the stream is finished?

**Reply by Liryna** [reply, marked as answer]
This will really depend on the application readying the file. If it lookup the file size at first and then sees the file size while readying is not the expected one...it might detect that as a failure.
It is really up to the media player implementation which is out of control.

**Reply by Liuyadi-Code** [reply]
I would like to ask when opening the file, in which function gets the data of the remote server?
